### ISSUE TITLE
Can't replicate different, multiple databases (event concurrency bug?)

### ISSUE THREAD

**Issue by cristiano-belloni** [orbitdb/orbit-db#258]
This is an issue I have while trying to replicate two different dbs (in my case a keyvalue and a feed):

```javascript
async function replicateFeed(address) {
    const feedDb = await orbitdb.feed(address)
    feedDb.events.on('replicated', () => console.log('feed replicated)', address)
    feedDb.events.on('replicate', () => console.log('feed replicate)', address)
    feedDb.events.on('replicate.progress', console.log('feed progress)', address)
}

async function replicateKvalue(address) {
    const feedDb = await orbitdb.keyvalue(address)
    feedDb.events.on('replicated', () => console.log('keyvalue replicated)', address)
    feedDb.events.on('replicate', () => console.log('keyvalue replicate)', address)
    feedDb.events.on('replicate.progress', console.log('keyvalue progress)', address)
}

await replicateKvalue(addressOfKvalue)
await replicateFeed(addressOfFeed)
```

If I run the above code, ~I get only the feed events. If I invert the bottom two lines, I get only the keyvalue events~ I get only the feed callback. The events, however, are sometimes the feed events and sometimes the keyvalue ones (ie, 50% of the time I get all the feeds and 50% all the keyvalues, despite the callbacks being called being only the feed one). If I invert the lines, vice versa.

Note that `replicateKvalue` and `replicateFeed` are **not** running concurrently: I `await` for each before running the other. What runs concurrently is the replication in the background. For some reason, only the last replication events seem to "win", but the updates are random. Looks like event rewriting?

In my case, waiting for replication serially (ie calling `replicateFeed` in the feed's `replicated` event handler) is not an option: I'm trying to write the "Twitter" demo and I'm doing these two replications for each followed user: If user are many, waiting serially for each of their databases and then waiting serially for each one of the users will take a lot of time.

**Comment by haadcode** [comment]
This is definitely a bug. I'll take a look asap. Thanks for reporting!

**Comment by haadcode** [comment]
@cristiano-belloni I'm having hard time to reproduce this. Can you share the full code for this? What are the exact addressed passed to each function? I've been looking into this but it doesn't happen here, I always get events from databases.

To clarify: **it should be possible to open and use multiple databases concurrently**.

**Comment by cristiano-belloni** [comment]
I'll try to simplify my code and reproduce this easily in the next days. Or
I can open my Gitlab PoC repo, if you like, but it's much harder to read.

I'm getting these results in Chrome, btw. I didn't try in Node.

**Comment by cristiano-belloni** [comment]
Anyway, that's my code. What I do is calling `enqueueDb` twice, once with an object like `{ type: 'user', dbAddress: aKeyValueAddress }` as the first param and the second time with an object like `{ type: 'feed', dbAddress: aFeedAddress }`. These are, respectively, the key-value user details and the feed of a followed user (which I created in another browser). LMK if you need the access to the repo.

```javascript
import IPFS from 'ipfs'
import OrbitDB from 'orbit-db'

let orbitdb
let access
let userDb
let feedDb
let followingDb

const replicateQueue = []
let isReplicating = false
let replicateCb

const ipfs = new IPFS({
  repo: '/vitriol-orbit/',
  EXPERIMENTAL: { pubsub: true }
})

function init() {
  return new Promise(res => {
    ipfs.on('ready', async () => {
      if (!orbitdb) orbitdb = new OrbitDB(ipfs)
      if (!access) access = {
        // Give write access to ourselves
        write: [orbitdb.key.getPublic('hex')]
      }
      res()
    })
  })
}

function getPublicKey() {
  if (orbitdb) return orbitdb.key.getPublic('hex')
  return null
}

export function getDbAddresses() {
  if (userDb && feedDb) return {
    userDb: userDb.address.toString(),
    feedDb: feedDb.address.toString()
  }
  return null
}

export async function onUserChange(onChange) {
  if (!orbitdb) await init()
  if (!userDb) userDb = await orbitdb.keyvalue('user-db', access)

  userDb.events.on('write', () => {
    onChange(getUser())
  })

  await userDb.load()
  onChange(getUser())
}

function getUser () {
  return {
    userName: userDb.get('userName'),
    firstName: userDb.get('firstName'),
    lastName: userDb.get('lastName'),
    publicKey: getPublicKey()
  }
}

export async function setUser(userName, firstName, lastName) {
  await userDb.put('userName', userName)
  await userDb.put('firstName', firstName)
  await userDb.put('lastName', lastName)
}

function getFeed() {
  return feedDb.iterator({limit: 10}).collect().reverse()
}

export async function onFeedChange(onChange) {
  if (!orbitdb) await init()
  if (!feedDb) feedDb = await orbitdb.feed('feed-db', access)

  feedDb.events.on('write', () => {
    onChange(getFeed())
  })

  await feedDb.load()
  onChange(getFeed())
}

export async function post(text) {
  feedDb.add({
    text,
    ts: Date.now()
  })
}

export async function onFollowingChange(onChange) {
  if (!orbitdb) await init()
  if (!followingDb) followingDb = await orbitdb.docstore('following-db', { indexBy: 'userDb' })

  followingDb.events.on('write', () => {
    onChange(getFollowing())
  })

  await followingDb.load()
  onChange(getFollowing())
}

function getFollowing() {
  return followingDb.query(() => true)
}

async function replicateUser(address) {
  console.log('$$ starting replicateUser', address)
  const userDb = await orbitdb.keyvalue(address)
  userDb.events.on('replicated', () => {
    console.log('**REPLICATED EVENT (^user^ has been replicated)', address)
    replicateCb({
      userName: userDb.get('userName'),
      firstName: userDb.get('firstName'),
      lastName: userDb.get('lastName'),
      publicKey: getPublicKey()
    }, {
      type: 'user',
      address
    })
  })
  userDb.events.on('replicate', () => {
    console.log('**REPLICATE EVENT (^user^ is being replicated)', address)
  })
  userDb.events.on('replicate.progress', (addr, hash, entry) => {
    console.log('**REPLICATE.PROGRESS EVENT (^user^ has progress)', entry.payload)
  })
  console.log('$$ ending replicateUser', address)
}

async function replicateFeed(address) {
  console.log('$$ starting replicateFeed', address)
  const feedDb = await orbitdb.feed(address)
  feedDb.events.on('replicated', () => {
    console.log('**REPLICATED EVENT (^feed^ has been replicated)', address)
    replicateCb(
      feedDb.iterator({limit: 10}).collect().reverse(),
      {
        type: 'feed',
        address
      }
    )
  })
  feedDb.events.on('replicate', () => {
    console.log('**REPLICATE EVENT (^feed^ is being replicated)', address)
  })
  feedDb.events.on('replicate.progress', (addr, hash, entry, progress, have) => {
    console.log('**REPLICATE.PROGRESS EVENT (^feed^ has progress)', entry.payload)
  })
  console.log('$$ ending replicateFeed', address)
}

async function replicate() {
  isReplicating = true
  const item = replicateQueue.splice(0, 1)[0]
  console.log('replicating', item)
  if (item) {
    const { type, dbAddress } = item
    switch(type) {
      case 'user':
        await replicateUser(dbAddress)
        await replicate()
        break
      case 'feed':
        await replicateFeed(dbAddress)
        await replicate()
        break
      default:
        break
    }
  } else {
    isReplicating = false
  }
}

export async function enqueueDb(dbInfo, cb) {
  console.log('enqueuing', dbInfo)
  if (cb) replicateCb = cb
  replicateQueue.push(dbInfo)
  if (!isReplicating) await replicate()
}

export async function follow(token) {
  return followingDb.put(token)
}
```

**Comment by cristiano-belloni** [comment]
Ok, the mystery is solved: I was using Firefox as the browser for the other user, and Firefox dies in the middle of replication with:

`TypeError: address is undefined orbit-db-address.js:17`

I guess that's the reason why the replication is not finished on the other side. If I use Chrome / incognito for the followed user, the replication seems to work. LMK if you want me to open a separate bug for FF.

I'm getting a lot of events, though. Like 12 events for `replicate` on the keyvalue db and 4 events for `replicated`. Is that ok? As usual, LMK if I need to open a separate bug for too many events.

**Comment by cristiano-belloni** [comment]
Sorry, it seems I wrote too soon. Now with the same code I'm in the situation of the first post, even with incognito mode. The behaviour seems erratic, I'm kinda confused.

**Comment by haadcode** [comment]
I think I figured it out. This is indeed a bug!

What's happening is, I believe, that in the example above, OrbitDB is using the same local data directory for both *peers*, ie. they write to the same directory and the data gets messed up in a way that the feedDb gets the userDb events. I didn't find yet what exactly causes this and where it happens, but definitely something we want to fix!

Meanwhile, can you try giving the databases a unique local directory with: `new OrbitDB(ipfs, <unique directory>)` and see if that fixes the problem on your end? Here's an example I used to debug this: https://gist.github.com/haadcode/4bba3ede25e4f2bca191f886e7564c3e

@cristiano-belloni let me know if this fixes it. I'll keep digging as to how to fix it.

**Comment by haadcode** [comment]
Spoke too as to what's causing it 😄 Found the cause: https://github.com/ipfs-shipyard/ipfs-pubsub-room/issues/25.

What's causing your problems @cristiano-belloni is that in our pubsub implementation, sometimes messages get handled in one topic and sometimes in the other and that's why you're seeing 50/50 chance for them to appear in one or the other.

I'll try to make a workaround fix for this today/tomorrow, but we'll need to fix it properly as discussed in the issue above.

**Comment by cristiano-belloni** [comment]
Thanks for your deep analysis @haadcode! If you need me to test something or run some branches, I'm happy to help :)

**Comment by haadcode** [comment]
Fix is coming soon! :)

**Comment by haadcode** [comment]
@cristiano-belloni this is now fixed in master (soon in npm as 0.18.3). Let me know if you still have the problem!

**Comment by haadcode** [comment]
Now published to npm as 0.18.3

**Comment by haadcode** [comment]
@cristiano-belloni we found out that the problem was actually in the message propagation layer in ipfs-pubsub-room. This was fixed by @pgte and the fixed version is now in [orbit-db v0.18.4](https://www.npmjs.com/package/orbit-db). 

Make sure to pull in the latest! 😄

**Comment by cristiano-belloni** [comment]
@haadcode I will, thanks (currently using your 18.3 workaround and it works,btw!)
